File: exercises/101-109_small_problems/medium_2/longest_sentence.rb
# Longest Sentence
# Write a program that reads the content of a text file and then prints the longest sentence in the file based on number of words.
# Sentences may end with periods (.), exclamation points (!), or question marks (?).
# Any sequence of characters that are not spaces or sentence-ending characters should be treated as a word.
# You should also print the number of words in the longest sentence.
#
# Input: A text file
# Output: Sentence with the most words
# Mental model: Given a text file, find the sentence with the most words.
# Return number of words in the longest sentence.

# Sentences end with . ! or ?

# Any sequence of characters that are not spaces or sentence ending characters should be treated as a word.
#
# File.open.each reads each line as ending with a \n
# Our sentences end with punctuation.
# Best way to do this: read file as a string, remove \n from string and then split on punctuation.
# Example:
# Text starts with:
# Four score and seven years ago our fathers brought forth
# on this continent a new nation, conceived in liberty, and
# dedicated to the proposition that all men are created
# equal. ....
# Script returns :
# The longest sentence is :
# It is rather for
# us to be here dedicated to the great task remaining
# before us -- that from these honored dead we take
# increased devotion to that cause for which they gave
# the last full measure of devotion -- that we here highly
# resolve that these dead shall not have died in vain
# -- that this nation, under God, shall have a new birth
# of freedom -- and that government of the people, by
# the people, for the people, shall not perish from the
# earth.
#
# and is 86 words long. (Note that each -- counts as a word.)
# #
# Data structures

# Input is a text file - can we use an array to store previous string + current string like with fibonacci
# or maybe store in a variable
#
# Algorithm
# Create new str lines
# Loop through text file
#   Add each line to string text without newlines

# Create new lines array
# Split string into lines array

# Loop through lines array
# Split each line into words
# Select line that has most words

# Return longest line
# Return longest line length

#
# Code

def get_longest_line(file)
  lines = ""

  File.open(file).each do |line|
    lines << line.gsub(/\n/, " ")
  end

  words = lines.split(/[.?!]/).map do |line|
    line.split
  end

  longest_line = words.sort_by! {|line| line.length }.last
end

def display_line(line)
  puts "The longest sentence is: "
  puts "#{line.join(" ")}"
  puts "and is #{line.length} words long."
end

def longest_line(file)
  longest_line = get_longest_line(file)

  display_line(longest_line)
end

longest_line("longest_sentence.txt")
longest_line("84-0.txt")

# Further Exploration
# You may have noticed that our solution fails to print the period at the end of the sentence. Can you write a solution that keeps the period printed at the end of each sentence?
#
# What about finding the longest paragraph or longest word? How would we go about solving that problem?

File: lessons/215_lessons/lesson_2/longestSentence.js
// Code Review: Longest Sentence
// Write a program that determines the sentence with the most words in some text. Sentences may end with periods (.), exclamation points (!), or question marks (?). Sentences always begin with a word character. You should treat any sequence of characters that are not spaces or sentence-ending characters, as a word. Log the longest sentence and its word count to the console. Pay attention to the expected output. Note that this problem is about manipulating and processing strings. As such, every detail about the string matters (e.g., case, punctuation, tabs, spaces, etc.).
//

// Input: A string
// Output: A string representing the sentence with the most words.
//
// Rules:
// - Sentences may end with periods (.), exclamation points (!) and question marks ?
// - Sequences of characters that are not spaces or sentence ending characters are words.
//
// - In each sentence, do we count characters like '--' as words? Yes.
// - Output contains the ending punctuation for each sentence. So while sentences may end with periods, exclamation points, and question marks, we have to keep those things in the sentence (which split doesn't do.)

// - Looking at the input, there is a space after each sentence ender (or a space beginning each sentence. Can we split on that? instead)

// Regex: split on space following .!?

// Mental model: Split string into sentences on the spaces following .!? Count words in each sentence. Return the sentence that has the most words.

// Data structurfe

// Algorithm
// 1. Split string into an array on .!?.
// 2. Sort sentences based on sentence length.
// 3. Return the sentence with the most words.
//
// To find sentence with the most words:
// 1. Split sentence into an array on spaces
// 2. Get length of array
//
// To sort sentences based on sentence length:
// 1. Pass in an anonymous function that returns a number (a - b) to get a number

// Example
let longText = 'Four score and seven years ago our fathers brought forth' +
  ' on this continent a new nation, conceived in liberty, and' +
  ' dedicated to the proposition that all men are created' +
  ' equal.' +
  ' Now we are engaged in a great civil war, testing whether' +
  ' that nation, or any nation so conceived and so dedicated,' +
  ' can long endure. We are met on a great battlefield of that' +
  ' war. We have come to dedicate a portion of that field, as' +
  ' a final resting place for those who here gave their lives' +
  ' that that nation might live. It is altogether fitting and' +
  ' proper that we should do this.' +
  ' But, in a larger sense, we can not dedicate, we can not' +
  ' consecrate, we can not hallow this ground. The brave' +
  ' men, living and dead, who struggled here, have' +
  ' consecrated it, far above our poor power to add or' +
  ' detract. The world will little note, nor long remember' +
  ' what we say here, but it can never forget what they' +
  ' did here. It is for us the living, rather, to be dedicated' +
  ' here to the unfinished work which they who fought' +
  ' here have thus far so nobly advanced. It is rather for' +
  ' us to be here dedicated to the great task remaining' +
  ' before us -- that from these honored dead we take' +
  ' increased devotion to that cause for which they gave' +
  ' the last full measure of devotion -- that we here highly' +
  ' resolve that these dead shall not have died in vain' +
  ' -- that this nation, under God, shall have a new birth' +
  ' of freedom -- and that government of the people, by' +
  ' the people, for the people, shall not perish from the' +
  ' earth.';

function longestSentence(text) {
  let sentences = text.split(/(?<=[.?!])\s/);
  let longest;

  sentences.sort(sortByLength);
  longest = sentences[0];
  
  console.log(longest);
  console.log('');
  console.log(`The longest sentence has ${getLength(longest)} words.`);
  console.log('');
}

function sortByLength(sentence1, sentence2) {
  return getLength(sentence2) - getLength(sentence1);
}

function getLength(sentence) {
  return sentence.split(/\s/).length;
}

longestSentence(longText);

longText = 'Four score and seven years ago our fathers brought forth' +
  ' on this continent a new nation, conceived in liberty, and' +
  ' dedicated to the proposition that all men are created' +
  ' equal.' +
  ' Now we are engaged in a great civil war, testing whether' +
  ' that nation, or any nation so conceived and so dedicated,' +
  ' can long endure. We are met on a great battlefield of that' +
  ' war. We have come to dedicate a portion of that field, as' +
  ' a final resting place for those who here gave their lives' +
  ' that that nation might live. It is altogether fitting and' +
  ' proper that we should do this.' +
  ' But, in a larger sense, we can not dedicate, we can not' +
  ' consecrate, we can not hallow this ground. The brave' +
  ' men, living and dead, who struggled here, have' +
  ' consecrated it, far above our poor power to add or' +
  ' detract. The world will little note, nor long remember' +
  ' what we say here, but it can never forget what they' +
  ' did here. It is for us the living, rather, to be dedicated' +
  ' here to the unfinished work which they who fought' +
  ' here have thus far so nobly advanced.'

  longestSentence(longText);

// console output
// It is rather for us to be here dedicated to the great task remaining before us -- that from these honored dead we take increased devotion to that cause for which they gave the last full measure of devotion -- that we here highly resolve that these dead shall not have died in vain -- that this nation, under God, shall have a new birth of freedom -- and that government of the people, by the people, for the people, shall not perish from the earth.
//
// The longest sentence has 86 words.


// Assuming the last sentence is removed:

// longestSentence(longText);

// console output
// Four score and seven years ago our fathers brought forth on this continent a new nation, conceived in liberty, and dedicated to the proposition that all men are created equal.
//
// The longest sentence has 30 words.
